### ISSUE TITLE
Subsequent calls to `fit` functions require resetting seeds, breaks reproducibility

### ISSUE THREAD

**Issue by gndowns** [fastai/fastai#2206]
#### Reproducibility and Seeding RNGs
The fastai [docs](https://docs.fast.ai/dev/test.html#getting-reproducible-results) describe how one can get reproducibility by seeding the various random number generators used. 

Unfortunately, after calling `fit_one_cycle` and loading a new model it appears the random number generators need to be _re_-seeded in order to get reproducible results. 

Does `fit_one_cycle` somehow reset the RNG seeds? Is this intended?

**Installation Details**
```
=== Software ===
python        : 3.7.3
fastai        : 1.0.52
fastprogress  : 0.1.21
torch         : 1.1.0
nvidia driver : 410.104
torch cuda    : 9.0.176 / is available
torch cudnn   : 7501 / is enabled

=== Hardware ===
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 11441MB | Tesla K80

=== Environment ===
platform      : Linux-4.15.0-1050-azure-x86_64-with-Ubuntu-16.04-xenial
distro        : #55-Ubuntu SMP Sat Jun 29 00:27:54 UTC 2019
conda env     : py35
python        : /home/gdowns/venv/bin/python
sys.path      : 
/opt/caffe/python
/opt/caffe2/build
/home/gdowns/venv/lib/python37.zip
/home/gdowns/venv/lib/python3.7
/home/gdowns/venv/lib/python3.7/lib-dynload
/usr/lib/python3.7
/home/gdowns/venv/lib/python3.7/site-packages
/usr/local/lib/python3.7/dist-packages
/usr/lib/python3/dist-packages
```

**To Reproduce**
The issue can be replicated using the code in this [gist](https://gist.github.com/gndowns/d4267f6e54c1c72ffc69e6a4b5baacbf)
When RNGs are seeded before each call to `fit_one_cycle`, all results are exactly the same. 
When RNGs are only seeded once at the beginning of the experiment, behaviour is unpredictable. 

**Sample Output of Gist**
```
Trial 1

epoch     train_loss  valid_loss  time
0         0.162521    0.064378    00:12
1         0.050993    0.031633    00:10
2         0.037767    0.028020    00:10


epoch     train_loss  valid_loss  time
0         0.166590    0.065955    00:10
1         0.048587    0.035391    00:10
2         0.032069    0.033567    00:10



Trial 2

epoch     train_loss  valid_loss  time
0         0.162521    0.064378    00:10
1         0.050993    0.031633    00:10
2         0.037767    0.028020    00:10


epoch     train_loss  valid_loss  time
0         0.162521    0.064378    00:10
1         0.050993    0.031633    00:10
2         0.037767    0.028020    00:10
```
**Expected behavior**
RNGs should only have to be seeded once at the beginning of a script in order to guarantee reproducible results. Calls to FastAI functions should not reset or re-seed RNGs.

**Comment by gndowns** [comment]
Further description of the problem for reference: https://forums.fast.ai/t/solved-reproducibility-where-is-the-randomness-coming-in/31628/14

**Comment by sgugger** [comment]
I'm not sure I understand what you think the bug is. In trial 1, you set the seeds before the first fit but not the second, so you get a different behavior in the second fit. A training is random by essence, since the data is shuffled at every epoch in the training set, so this is completely expected.

In trial 2 you set the seeds before each call so get the same behavior each time. This all seems perfectly normal to me.

**Comment by gndowns** [comment]
Sorry maybe I wasn't clear. I do not understand why seeds should have to be set before each call to fit. 

If the generators are seeded once at the beginning of the session, why should they ever have to be seeded again? What is causing the seeds to be reset?

**Comment by sgugger** [comment]
In trial 1 you set the seed and ask for n random numbers (first training) then ask for n new random numbers, which are going to be random since you didn't reset the seed (second training).

In trial 2 you set the seed and ask for n random numbers (first training) then set the seed again and ask for n random numbers, which are going to be the same ones as before (second training).

Setting the seed just initialize your random number generators, then, when you draw numbers you need to set the seed again to get the same ones.

**Comment by gndowns** [comment]
Sorry see my edit above for an example of expected behaviour. Why does calling `fit_one_cycle` reset the RNGs?

**Comment by sgugger** [comment]
Your expected behavior is the weird one IMO. You say calling `fit_one_cycle` resets the RNGs but it does not, that is exactly what I'm explaining in my previous posts. The RNGs have been set with a seed, you ask for random permutations three times in a row with your three epochs, then three permutations again which are going to be
- the same as before if you set the random seed again (which would reset the RNGs)
- not the same as before if you don't set the seed again

In any case there is nothing in the fastai library or PyTorch that induces this behavior since it's the normal behavior of the random libraries. There is probably something in Keras/TF to make it behave this way on the other hand.

Closing the issue since there is no bug here.

**Comment by gndowns** [comment]
@sgugger Ah I see now, my misunderstanding sorry. Thanks for the explanation.
